host/xtest/pkcs11_1000.c

```diff
@@ -10021,3 +10021,386 @@ static void xtest_pkcs11_test_1030(ADBG_Case_t *c)
+	CK_BBOOL g_sign = CK_FALSE;
+	CK_BBOOL g_sign_recover = CK_FALSE;
+	CK_BBOOL g_unwrap = CK_FALSE;
+	CK_BBOOL g_extract = CK_FALSE;
+	CK_BBOOL g_asensitive = CK_FALSE;
+	CK_BBOOL g_nextract = CK_FALSE;
+	CK_BBOOL g_wrap_with_trusted = CK_FALSE;
+	CK_BBOOL g_always_authenticate = CK_FALSE;
+
+	/* Note: Tests below expects specific order of elements */
+	CK_ATTRIBUTE get_public_template[] = {
+		{ CKA_CLASS, &g_class, sizeof(g_class) },
+		{ CKA_KEY_TYPE,	&g_key_type, sizeof(g_key_type) },
+		{ CKA_ID, g_id, sizeof(g_id) },
+		{ CKA_DERIVE, &g_derive, sizeof(g_derive) },
+		{ CKA_LOCAL, &g_local, sizeof(g_local) },
+		{ CKA_KEY_GEN_MECHANISM, &g_keygen_mecha, sizeof(g_keygen_mecha) },
+		{ CKA_SUBJECT, g_subject, sizeof(g_subject) },
+		{ CKA_ENCRYPT, &g_encrypt, sizeof(g_encrypt) },
+		{ CKA_VERIFY, &g_verify, sizeof(g_verify) },
+		{ CKA_VERIFY_RECOVER, &g_verify_recover, sizeof(g_verify_recover) },
+		{ CKA_WRAP, &g_wrap, sizeof(g_wrap) },
+		{ CKA_TRUSTED, &g_trusted, sizeof(g_trusted) },
+		{ CKA_PUBLIC_KEY_INFO, g_public_key_info, sizeof(g_public_key_info) },
+	};
+
+	/* Note: Tests below expects specific order of elements */
+	CK_ATTRIBUTE get_private_template[] = {
+		{ CKA_CLASS, &g_class, sizeof(g_class) },
+		{ CKA_KEY_TYPE,	&g_key_type, sizeof(g_key_type) },
+		{ CKA_ID, g_id, sizeof(g_id) },
+		{ CKA_DERIVE, &g_derive, sizeof(g_derive) },
+		{ CKA_LOCAL, &g_local, sizeof(g_local) },
+		{ CKA_KEY_GEN_MECHANISM, &g_keygen_mecha, sizeof(g_keygen_mecha) },
+		{ CKA_SUBJECT, g_subject, sizeof(g_subject) },
+		{ CKA_SENSITIVE, &g_sensitive, sizeof(g_sensitive) },
+		{ CKA_DECRYPT, &g_decrypt, sizeof(g_decrypt) },
+		{ CKA_SIGN, &g_sign, sizeof(g_sign) },
+		{ CKA_SIGN_RECOVER, &g_sign_recover, sizeof(g_sign_recover) },
+		{ CKA_UNWRAP, &g_unwrap, sizeof(g_unwrap) },
+		{ CKA_EXTRACTABLE, &g_extract, sizeof(g_extract) },
+		{ CKA_ALWAYS_SENSITIVE, &g_asensitive, sizeof(g_asensitive) },
+		{ CKA_NEVER_EXTRACTABLE, &g_nextract, sizeof(g_nextract) },
+		{ CKA_WRAP_WITH_TRUSTED, &g_wrap_with_trusted, sizeof(g_wrap_with_trusted) },
+		{ CKA_ALWAYS_AUTHENTICATE, &g_always_authenticate, sizeof(g_always_authenticate) },
+		{ CKA_PUBLIC_KEY_INFO, g_public_key_info, sizeof(g_public_key_info) },
+	};
+
+	CK_BYTE in_data[1024] = { 0 };
+	CK_ULONG in_data_size = 0;
+	CK_BYTE signature[1024] = { 0 };
+	CK_ULONG signature_len = 0;
+
+	Do_ADBG_BeginSubCase(c, "Test CKM_RSA_X_509 %u - Sign/Verify", rsa_bits);
+
+	Do_ADBG_BeginSubCase(c, "Generate key pair and data");
+
+	modulus_bits = rsa_bits;
+	rv = C_GenerateKeyPair(session, &mechanism, public_key_template,
+			       ARRAY_SIZE(public_key_template),
+			       private_key_template,
+			       ARRAY_SIZE(private_key_template),
+			       &public_key, &private_key);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto err;
+
+	/*
+	 * Current implementation of the PKCS#11 TA requires that the
+	 * message to sign has to size of the private key. There is
+	 * no constraint regarding the padding scheme. Lets's use a
```

Comment:
This is wrong: there is 1 constraint: the message shall represent a value smaller than the key modulus. We need to at least clear the leading bit (`in_data[0] &= 0x7f;`) top achieve that.

By the way, typo above ("has **~to~ the** size of the private key.")